File: 01/solver2.exs
#!/usr/bin/elixir

{:ok, file} = File.read("input.txt")
lst = file |> String.trim
           |> String.split("\n")
           |> Enum.map(&String.to_integer/1)
           |> Enum.sort
# lst is now a list with sorted integers

defmodule Pivoting do
  # Just exit when greater than goal (list is sorted):
  def deep_search(_list, p1, p2, p3) when p1 + p2 + p3 > 2020, do: nil

  # Goal reached?
  def deep_search(_list, p1, p2, p3) when p1 + p2 + p3 == 2020 do
    IO.puts("p1: #{p1}, p2: #{p2}, p3: #{p3} => p1 * p2 * p3 = #{p1 * p2 * p3}")
  end

  # Do the actual search:
  def deep_search(list, p1, p2, p3) when p1 + p2 + p3 < 2020 do
    # Go deep for p2 (move p3; don't go if list is shorter):
    if(Enum.at(list, 3, false), do: Pivoting.deep_search((tl list), p1, p2, Enum.at(list, 3, 0)), else: list)

    # Go deep for p1 (move p2 & p3; don't go if list is shorter):
    if(Enum.at(list, 4, false), do: Pivoting.deep_search((tl list), p1, Enum.at(list, 3, 0), Enum.at(list, 4, 0)), else: list)
  end
end

Pivoting.deep_search(lst, (hd lst), Enum.at(lst, 1, 0), Enum.at(lst, 2, 0))

File: 01/solver2.rb
#!/usr/bin/env ruby

ary = File.read("input.txt").split("\n").map(&:to_i).sort

def pivoting(list, p1, p2, p3)
  # Guard & exit if greater than goal (list is sorted):
  return if p1 + p2 + p3 > 2020

  # Output when goal is reached:
  if p1 + p2 + p3 == 2020
    puts "p1: #{p1}, p2: #{p2}, p3: #{p3} => p1 * p2 * p3 = #{p1 * p2 * p3}"
    return
  end

  # Go deep for p2 (move p3):
  pivoting(list[1..-1].dup, p1, p2, list[3]) if list.size > 3

  # Go deep for p1 (move p2 & p3):
  pivoting(list[1..-1].dup, p1, list[3], list[4]) if list.size > 4
end

pivoting(ary, ary[0], ary[1], ary[2])
